Add check to prevent NumPy ndarray from being treated as tuple when indexing

pac48 opened the pull request:
Fixes #108689

pac48 committed:
change heuristic value for deciding whether to convert to tuple

Signed-off-by: Paul Gesel <paulgesel@gmail.com>

pac48 committed:
add test for multidimensional numpy array indexing

Signed-off-by: Paul Gesel <paulgesel@gmail.com>

lezcano reviewed (changes requested):
A better fix would be to detect whether the input is an ndarray and return false in that case. The current proposal may silently break existing code, so that's a no-no.

In general, using mixing NumPy and PyTorch arrays, when using PyTorch ops, is not supported. If you want to mix them, bring the NumPy ndarray to PyTorch using `torch.from_numpy`, which uses the buffer protocol and doesn't copy.

pac48 committed:
revert heuristic value change. Check for ndarray instead

Signed-off-by: Paul Gesel <paulgesel@gmail.com>

pac48 commented:
> A better fix would be to detect whether the input is an ndarray and return false in that case. The current proposal may silently break existing code, so that's a no-no.
> 
> In general, using mixing NumPy and PyTorch arrays, when using PyTorch ops, is not supported. If you want to mix them, bring the NumPy ndarray to PyTorch using `torch.from_numpy`, which uses the buffer protocol and doesn't copy.

Thank you for the helpful feedback. I reverted the heuristic value change and added a check using `PyArray_CheckExact` to detect a NumPy ndarray.

pac48 committed:
Merge branch 'main' into fix-numpy-tensor-indexing

lezcano commented on torch/csrc/autograd/python_variable_indexing.cpp:
```diff
@@ -291,6 +292,9 @@ static inline bool treatSequenceAsTuple(PyObject* index) {
   if (n >= 32) {
     return false;
   }
+  if (PyArray_CheckExact(index)) {
+    return false;
+  }
```
You cannot do this unconditionally, as NumPy is not a runtime dependency of PyTorch. You need to do this just if `USE_NUMPY` is true. You also need to initialize NumPy, by checking that
```python
  TORCH_CHECK(is_numpy_available(), "Numpy is not available");
```
as it may not be initialised, in which case `CheckExact` may break.

Also, could you move this up, next to the tensor check and update the  comment to note that we also consider NumPy arrays?

pac48 committed:
add TORCH_CHECK for is_numpy_available

Signed-off-by: Paul Gesel <paulgesel@gmail.com>

lezcano reviewed (commented):
Just a minor point

lezcano commented on torch/csrc/autograd/python_variable_indexing.cpp:
```diff
@@ -270,6 +272,12 @@ static inline bool treatSequenceAsTuple(PyObject* index) {
     return false;
   }
+#ifdef USE_NUMPY
+  TORCH_CHECK(::torch::utils::is_numpy_available(), "Numpy is not available");
+  if (PyArray_CheckExact(index)) {
```
Sorry, I edited the message. I think this should be
```suggestion
  if (::torch::utils::is_numpy_available() && PyArray_CheckExact(index)) {
```

pac48 committed:
Update torch/csrc/autograd/python_variable_indexing.cpp

Co-authored-by: Mario Lezcano Casado <3291265+lezcano@users.noreply.github.com>

lezcano reviewed (approved):
Cool! Thank you for the fix!

lezcano commented:
@pytorchbot merge

pac48 commented:
> Just a minor point

Thanks for the suggestion. As far as the comment explaining numpy support, I was thinking something along the lines of:
```c++
  //  Allow indexing with ndarray if numpy compilation is enabled. An ndarray
  //  index should not be treated as a tuple since the indexing has a different 
  //  syntax.
```

pac48 committed:
update comment about ndarray indexing

Signed-off-by: Paul Gesel <paulgesel@gmail.com>

lezcano commented:
@pytorchbot merge
